src/outputs/parthenon_hdf5.cpp

```diff
@@ -192,436 +365,652 @@ void PHDF5Output::genXDMF(std::string hdfFile, Mesh *pm, SimTime *tm) {
+      auto ci = get_MeshBlockDataIterator(pmb);
+      for (auto &v : ci.vars) {
+        // Note index l transposed to interior
+        if (var_name.compare(v->label()) == 0) {
+          auto v_h = v->data.GetHostMirrorAndCopy();
+          for (int k = out_kb.s; k <= out_kb.e; ++k) {
+            for (int j = out_jb.s; j <= out_jb.e; ++j) {
+              for (int i = out_ib.s; i <= out_ib.e; ++i) {
+                for (int l = 0; l < vlen; ++l) {
+                  tmpData[index++] = static_cast<OutT>(v_h(l, k, j, i));
+                }
+              }
+            }
           }
+
+          found = true;
+          break;
         }
       }
-      // write dataset to file
-      WRITEH5SLAB2(vWriteName.c_str(), tmpData.data(), file, local_start, local_count,
-                   vLocalSpace, vGlobalSpace, property_list);
+
+      if (vinfo.is_sparse) {
+        size_t sparse_idx = sparse_field_idx.at(vinfo.label);
+        sparse_expanded[b_idx * num_sparse + sparse_idx] = found;
+      }
+
+      if (!found) {
+        if (vinfo.is_sparse) {
+          hsize_t N = varSize * vlen;
+          memset(tmpData.data() + index, 0, N * sizeof(OutT));
+          index += N;
+        } else {
+          std::stringstream msg;
+          msg << "### ERROR: Unable to find dense variable " << var_name << std::endl;
+          PARTHENON_FAIL(msg);
+        }
+      } else {
+        found_any = true;
+      }
+    }
+
+    // If found_any is true, it means that at least one local block has data for the
+    // current variable writing, so we write the tmpData buffer to the HDF5 file. Note,
+    // the tmpData buffer may contain some 0's for the local blocks that don't have this
+    // variable. It's ok to write these 0's because compression will take care of them.
+    // Otherwise, if found_any is false, then none of the local blocks have data for this
+    // variable, so we don't need to write a buffer of all 0's.
+    if (found_any) {
+      // write data to file
+      HDF5WriteND(file, var_name, tmpData.data(), 5, p_loc_offset, p_loc_cnt, p_glob_cnt,
+                  pl_xfer, pl_dcreate);
     }
   }
 
-  // generate XDMF companion file
-  (void)genXDMF(filename, pm, tm);
+  // NumVariables - number of variables within each dataset
+  std::vector<size_t> numVariables;
+  // VariablesNames - Names of variables within each dataset
+  std::vector<std::string> variableNames;
+
+  for (const auto &vi : all_unique_vars) {
+    const auto &component_labels = vi.component_labels;
+
+    if (component_labels.size() > 0) {
+      numVariables.push_back(component_labels.size());
+      for (const auto &label : component_labels) {
+        variableNames.push_back(label);
+      }
+    } else {
+      numVariables.push_back(1);
+      variableNames.push_back(vi.label);
+    }
+  }
+
+  HDF5WriteAttribute("NumVariables", numVariables, info_group);
+  HDF5WriteAttribute("VariableNames", variableNames, info_group);
```

Comment:
```suggestion
  // NumComponents - number of components within each dataset
  std::vector<size_t> numComponents;
  // ComponentNames - Names of components within each dataset
  std::vector<std::string> componentNames;

  for (const auto &vi : all_unique_vars) {
    const auto &component_labels = vi.component_labels;

    if (component_labels.size() > 0) {
      numComponents.push_back(component_labels.size());
      for (const auto &label : component_labels) {
        componentNames.push_back(label);
      }
    } else {
      numComponents.push_back(1);
      componentNames.push_back(vi.label);
    }
  }

  HDF5WriteAttribute("NumComponents", numComponents, info_group);
  HDF5WriteAttribute("ComponentNames", componentNames, info_group);
```
This naming convention I introduced is a bit confusing, especially `VariableNames`. What do you think about renaming these as "Components" as they're named in the metadata and elsewhere in the code?

Reply:
Good idea. If you don't mind, I think I'll change it to `num_components` and `component_names` instead to go with the snake case style used in that file

Reply:
For naming the C++ variables? Yes, you should do that.

Reply:
Done. Also updated gold to v7 and yes, the name in the HDF5 metadata changed